docs/ocis/adr/0015-events.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "15. oCIS Event System"
+weight: 15
+date: 2022-02-01T12:56:53+01:00
+geekdocRepo: https://github.com/owncloud/ocis
+geekdocEditPath: edit/master/docs/ocis/adr
+geekdocFilePath: 0015-events.md
+---
+
+* Status: proposed
+* Deciders: @butonic, @micbar, @dragotin, @c0rby, @wkloucek
+* Date: 2022-01-21
+
+## Context and Problem Statement
+
+### Overview
+
+To be able to implement simple, flexible and independant inter server communication there is the idea to implement an event system in oCIS. A service can send out events which are received by one or more other services. The receiving service can cause different kind of actions based on the event by utilizing the information that the event carries.
+
+###  Example: EMail Notification
+
+A simple example is the notification feature for oCIS: Users should receive an email when another user shares a file with them. The information, that the file was shared should go out as an event from the share manager, carrying the information which file was shared to which receiver. A potential notification service that really sends out the email listens to these kind of events and sends the email out once on every received event of that specific type.
+
+## Decision Drivers
+
+- Events are supposed to simplify things and raise flexibility, also considering extensions that are not directly controlled by the ownCloud project.
+- Events should bring flexibility in the implementation of sending and receiving services.
+- Events should not obsolete other mechanisms to communicate, ie. grpc calls.
+- Sending an event has to be as little resource consuming for the sender as possible.
+- Events are never user visible.
+
+## Considered Options
+
+1. Lightweight Events with Event Queue and "At most once" QoS
+2. As 1., but with "At least once" QoS
+
+## Options
+
+### 1. Lightweight Events with Event Queue and "At most once" QoS
+
+Reva will get a messaging service that is available to all services within oCIS and Reva. It is considered as one of the mandatory services of the oCIS system. If the messaging backend is not running, neither Reva nor oCIS can be considered healthy and should shut down.
```

Comment:
I'm not sure about this one. Read my other comment first then come back here.

As far as I understood the discussions our event bus really is just a way to notify other services of events. While delivery is not guaranteed it can be used by the storage providers to notify other services that a file finished uploading or that a rename happened. The thumbnailer and the indexer can react to these events immediately.

For our uploads we do need to keep track of several stages of an upload:
0. upload in progress
1. all bytes arrived
2. preprocessing like anti virus scanning or confidentiality check
3. making file avalable for download
4. postprocessing like indexing and thumbnailing
A similar workflow also applies to file renames and we agreed to use the saga pattern, as we already have a file that reflects the upload metadata. It can be used to persist the state. The at most once delivery event bus is then used to notify other services.

Now I acutally think that ... the event bus does not have to be made highly available in this case. What is the point if you cannot guarantee delivery anyway? I think this is a pro for at-most-once. Or put the other way around: if we rely on at-least-once, yes, then the event bus is a mandatory service and ocis / reva can defacto not work without it. Which is a con for an at-most-once event bus.

Reply:
ok, i get the argument. we have both options described and the pros and cons reflect what you think.